Improve sample shape plotting performance

DannyHindson opened the pull request:
**Description of work.**

Speed up performance of the plotting of the sample shape and container in cases where shapes are complex. Normally anything involving mesh shapes is the slowest type of shape but I've found that complex CSG shapes can cause problems with the shape plotting. This is because the CSG shapes need converting to a mesh before plotting which involves a costly call to a third party library called Open Cascade.

The performance improvement has been achieved by reducing the number of calls to getMesh from the sample shape plotting code.

The reason I noticed this was a bit of investigation I was doing for one of the ENGINX instrument scientists which involved me creating a description of a collimator from lots of CSG cuboids. The getMesh call in this case involves creating a mesh description of each CSG shape.

**To test:**

Run this script and then "Show Sample shape" from the right click menu of the ws workspace. Verify it's quicker than before. For me it still takes ~30 seconds so not rapid but better than it was:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = Load("ENGINX00305761") # 2513 histograms, 10186 bins

cuboid_height=0.85
cuboid_depth=0.0045246
cuboid_width=0.2
z_min = -0.364503 # -0.364503 to 0.367428
z_spacing = 0.0045746
n_cuboids=161 # giving 160 veins

cuboid_xml_template = f" \
<cuboid id='cuboid-1'> \
<height val='{cuboid_height}'  /> \
<width val='{cuboid_width}' />  \
<depth  val='{cuboid_depth}' />  \
<centre x='0.5' y='0.0' z='-0.1'  />  \
</cuboid>"

multiple_cuboid_xml = ''
algebra_xml = "<algebra val='"
for i in range(1,n_cuboids):
    single_cuboid_xml = cuboid_xml_template.replace("cuboid-1","cuboid-" + str(i))
    single_cuboid_xml = single_cuboid_xml.replace("z='-0.1'","z='" + str(z_min + i*z_spacing) + "'")
    multiple_cuboid_xml += single_cuboid_xml
    if (i==1):
        algebra_xml += "cuboid-" + str(i)
    else:
        algebra_xml += " : cuboid-" + str(i)
algebra_xml += "'/>"
multiple_cuboid_xml = multiple_cuboid_xml + algebra_xml

SetSample(ws,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,
                  'Radius': 2.0, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'Fe'},
          ContainerGeometry={'Shape': 'CSG', 'Value': multiple_cuboid_xml},
          ContainerMaterial={'ChemicalFormula': 'V'})
```

*There is no associated issue.*

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Reduce the number of calls to getMesh

Speed up performance of showing sample shape and container
in cases where shapes are complex

DannyHindson committed:
Adjust tests to use new get mesh functions

jhaigh0 commented on qt/python/mantidqt/mantidqt/plotting/sample_shape.py:
```diff
@@ -241,9 +244,8 @@ def get_component_shape_from_workspace(workspace_with_components, component_inde
-    if shape:
-        mesh = shape.getMesh()
+    mesh = get_valid_sample_mesh_from_workspace(workspace)
+    if mesh is not None:
```
I think `if mesh` is equivalent to `if mesh is not None` but okay if you just want to be explicit about it 👍

jhaigh0 commented on qt/python/mantidqt/mantidqt/plotting/test/test_sample_shape.py:
```diff
@@ -116,42 +116,42 @@ def tearDown(self) -> None:
         workspace = setup_workspace_shape_from_CSG_merged()
         self.assertTrue(workspace.sample().getShape())
-        self.assertTrue(sample_shape.get_valid_sample_shape_from_workspace(workspace))
+        self.assertTrue(sample_shape.get_valid_sample_mesh_from_workspace(workspace) is not None)
```
You could use `self.assertIsNotNone()` for these if you like.

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIsNotNone

jhaigh0 commented on qt/python/mantidqt/mantidqt/plotting/test/test_sample_shape.py:
```diff
@@ -116,42 +116,42 @@ def tearDown(self) -> None:
         workspace = setup_workspace_sample_container_and_components_from_mesh()
         mock_is_mesh_not_empty.return_value = False
-        self.assertFalse(sample_shape.get_valid_container_shape_from_workspace(workspace))
+        self.assertFalse(sample_shape.get_valid_container_mesh_from_workspace(workspace))
```
`self.assertIsNone()` would be more readable in these cases.

jhaigh0 reviewed (changes requested):
Nice change to localise those `getMesh()` calls. Also took around 30 seconds or so to show the shape for me, I would be a nice if there was a progress indicator for that I think (not in this pr though).

Just some minor code review comments.

DannyHindson committed:
Use more readable assert statements

DannyHindson committed:
Remove further duplicated calls to getMesh

Also put proper test in for set_axes_to_largest_mesh. It was
previously a duplicate of test_get_overall_limits_for_sample_only

DannyHindson commented:
I decided in the end that it was a bit rubbish that the view still takes 30-40 seconds to load. So I've just pushed a change that properly speeds this up by reducing the number of calls to getMesh even further. Each shape should now just be retrieved once with a single call to getMesh. For me it now takes about 10 seconds.

jhaigh0 reviewed (commented):
New changes work great, even quicker than before ⚡. I had a spinny wheel cursor this time when it was loading which is good enough feedback now the time to load is much smaller.

I've added one reformatting suggestion to get rid of the raw loop from `create_plot` but other than that all seems good. Much better to pass the meshes about rather than creating them in every function.

DannyHindson committed:
Pull out logic to get component meshes into separate function

jhaigh0 reviewed (changes requested):
Ah, Almost forgot.... I think this should have a release note since the users should notice the difference.

DannyHindson committed:
Add release note
